Docs: improve documentation of no-return-await (#13215)
eslint/eslint: docs/rules/no-return-await.md

# Disallows unnecessary `return await` (no-return-await)

> Using `return await` inside an `async function`keeps the current function in the call stack until the Promise that is being awaited have resolved. You can take a shortcut and just return the Promise right away to avoid this, saving an extra microtask before resolving the overarching Promise.

Reviewer: Could we incorporate this other caveat in this documentation?
```suggestion
Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited have resolved. You can take a shortcut and just return the Promise right away to avoid this, saving an extra microtask before resolving the overarching Promise. `return await` can also be  used in a try/catch statement to catch errors from another Promise-based function.
```

Author: Of course, I forgot to add that back after rewriting the paragraph. Thanks!

> Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited have resolved. You can take a shortcut and just return the Promise right away to avoid this, saving an extra microtask before resolving the overarching Promise. `return await` can also be  used in a try/catch statement to catch errors from another Promise-based function.

Reviewer: ```suggestion
Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited has resolved, at the cost of an extra microtask before resolving the outer Promise. `return await` can also be used in a try/catch statement to catch errors from another function that returns a Promise.
```

> The only visible change when doing this is that the function will no longer be a part of the stack trace, if an error is thrown asyncrously from the Promise being returned.

Reviewer: ```suggestion
The only visible change when doing this is that the function will no longer be a part of the stack trace, if an error is thrown asynchronously from the Promise being returned.
```

> The only visible change when doing this is that the function will no longer be a part of the stack trace, if an error is thrown asynchronously from the Promise being returned.

Reviewer: ```suggestion
You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.
```

## When Not To Use It

There are a few reasons you might want to turn this rule off:

- If you want to use `await` to denote a value that is a thenable
- If you do not want the performance benefit of avoiding `return await`

> - If you still want the functions to show up in stack traces

Reviewer: ```suggestion
- If you want the functions to show up in stack traces (useful for debugging purposes)
```
